src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs

```diff
@@ -175,25 +175,48 @@ private static bool UserDefinedConversionIsAllowed(ExpressionSyntax expression,
                 }
 
                 var parameter = argument.DetermineParameter(semanticModel, cancellationToken: cancellationToken);
-                if (parameter != null && parameter.IsParams)
+                return ParameterTypeMatchesParamsElementType(parameter, castType, semanticModel);
+            }
+
+            if (parent is AttributeArgumentSyntax attributeArgument)
+            {
+                if (attributeArgument.Parent is AttributeArgumentListSyntax attributeArgumentList)
                 {
-                    Debug.Assert(parameter.Type is IArrayTypeSymbol);
+                    // We don't check the position of the argument because in attributes it is allowed that 
+                    // params parameter are positioned in between if named arguments are used.
+                    // The *single* argument check above is also broken: https://github.com/dotnet/roslyn/issues/20742
```

Comment:
At this position we need to find the parameter that fits to the argument (DetermineParameter). We pass this information to ParameterTypeMatchesParamsElementType which is only interested in `params` parameters. We could avoid the call to DetermineParameter if we are sure that looking at the `attributeArgumentList` and the `attributeArgument` in it we can be sure that it

1. can't be a `params` parameter or
2. if it might be a `params` parameter the argument would be a single element array passed to the parameter.

I don't think it is possible to determine either of that conditions by only looking at the `attributeArgumentList` and the `attributeArgument`.

Reply:
I was specifically referring to how this:

> If there are any arguments to the right, we can assume that this is not a *single* argument passed to a params parameter.

Relates to this:

> The *single* argument check above is also broken

It seems that you could identify a single argument as an argument meeting both of the following conditions:

1. The argument appears first in the list
2. The next argument, if any, has a non-null `NameEquals` property, indicating that it is not an argument

Reply:
I was specifically referring to how this:

> If there are any arguments to the right, we can assume that this is not a *single* argument passed to a params parameter.

Relates to this:

> The *single* argument check above is also broken

It seems that you could identify a single argument as an argument meeting both of the following conditions:

1. The argument appears first in the list
2. The next argument, if any, has a non-null `NameEquals` property, indicating that it sets a named property instead of a parameter

Reply:
➡️ Filed #25790 as a follow-up

Reply:
The *single argument* comment is misleading. It's really asking for a *single argument passed to a params parameter*.